src/completion.rs

```diff
@@ -365,12 +365,33 @@ trait Completable {
 
 // Reports if the needle has a fuzzy match with the haystack.
 //
-// It is assumed that the haystack is the name of an identifier and the needle is a partial
-// identifier.
+// This match follows this logic: if the needle is sufficiently small,
+// check to see if the haystack starts with the needle *or* if the needle
+// is found as a camel case word in the haystack. If the needle looks like
+// a full word or more than one word, find the needle, case-insensitively, in
+// the haystack.
+#[allow(clippy::unwrap_used)]
 fn fuzzy_match(haystack: &str, needle: &str) -> bool {
-    return haystack
-        .to_lowercase()
-        .contains(needle.to_lowercase().as_str());
+    // This number seems "magic," and maybe it is, but it seemed a good threshold for "are you
+    // typing a word?"
+    if needle.len() > 3 {
```

Comment:
I think language server completion does not refresh the completions as one continues to type unless `isIncomplete: true` is specified in the result of the completion. So effectively only the second branch would do the matching on the server side unless we also change `isIncomplete` (when typing an identifier from scratch).

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionList

Reply:
Yeah, this is the real crux of whether this patch should go in. I wonder if `isComplete` is doing too much work, and that's why the behavior was the way it was before. "Hey, here's everything that could potentially match no matter what you type next. You figure out what to show." So it's either "be easy on the server side but put a lot of work on the client" or "make a lot of requests to the server so that the client doesn't have to be too smart."

I think my default is to encapsulate as much logic in the server as possible, and let the client be as dumb as it can. I'm not sure that's what the spec intended; in fact, it seems like the spec intends for the client to be much smarter than the server. Lemme give it a think, I guess.

Reply:
Yeah, the server vs client responsibility isn't always so clear. I think LSP generally does a lot on the server side as each command mostly implements one specific feature on the client side.

(I think intellij has a different model where the server integrates more deeply into the client, such that the client has direct AST access etc).